topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,183 @@
+
+As LSP became available in the 2023.2 EAP7 of IntelliJ-based IDEs, the plugin must target IntelliJ IDEA Ultimate `2023.2` or later.
+
+Example <path>build.gradle.kts</path> configuration:
+
+```kotlin
+plugins {
+  // ...
+  id("org.jetbrains.intellij") version "1.15.0"
+}
+
+intellij {
+  version = "232-EAP-SNAPSHOT"
+  type = "IU"
+}
+```
+
+For projects based on the [](plugin_github_template.md), update the Gradle IntelliJ Plugin to the latest version, and amend the <path>gradle.properties</path> file as follows:
+
+```properties
+platformType = IU
+platformVersion = 232-EAP-SNAPSHOT
+pluginSinceBuild = 232
+pluginUntilBuild = 232.*
+```
+
+The <path>plugin.xml</path> configuration file needs to specify the dependency on the IntelliJ IDEA Ultimate module:
+
+```xml
+<idea-plugin>
+   <!-- ... -->
+   <depends>com.intellij.modules.platform</depends>
+   <depends>com.intellij.modules.ultimate</depends>
+</idea-plugin>
+```
+
+The LSP API sources are bundled in IntelliJ IDEA Ultimate and can be found within the <path>[IDE]/lib/src/src_lsp-openapi.zip</path> archive.
+
+## Supported features
+
+The initial LSP support within the IntelliJ Platform covers the following features:
+
+- Errors/warnings highlighting ([textDocument/publishDiagnostics](https://microsoft.github.io/language-server-protocol/specification/#textDocument_publishDiagnostics))
+- Quick-fixes for these errors/warnings ([textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specification/#textDocument_codeAction))
+- Code completion ([textDocument/completion](https://microsoft.github.io/language-server-protocol/specification/#textDocument_completion))
+- Go to Declaration ([textDocument/definition](https://microsoft.github.io/language-server-protocol/specification/#textDocument_definition))
+- More features will be announced soon.
+
+## Basic implementation
+
+A minimal LSP integration must implement `LspServerSupportProvider` along with a service descriptor and register it as a `com.intellij.platform.lsp.serverSupportProvider` extension point:
+
+```kotlin
+import com.intellij.platform.lsp.api.LspServerSupportProvider
+import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor
+
+class FooLspServerSupportProvider : LspServerSupportProvider {
+  override fun fileOpened(project: Project, file: VirtualFile, serverStarter: LspServerStarter) {
+    if (file.extension == "foo") {
+      serverStarter.ensureServerStarted(FooLspServerDescriptor(project))
+    }
+  }
+}
+
+private class FooLspServerDescriptor(project: Project) : ProjectWideLspServerDescriptor(project, "Foo") {
+  override fun isSupportedFile(file: VirtualFile) = file.extension == "foo"
+  override fun createCommandLine() = GeneralCommandLine("foo", "--stdio")
+}
+```
+
+After providing the implementation of the LSP Server support provider extension point, register it in the <path>plugin.xml</path> file as follows:
```

Comment:
I think there is no need for this paragraph and the following snippet.

Reply:
Does it differ from a regular extension registration? We describe registering tens of extensions in the documentation, and it's enough to just mention what should be registered in what EP. There is no need for code snippets. Also, this one doesn't show registration, but also adding dependencies (so it's not consistent with the introducing text), what was mentioned on the page already.

Reply:
Removed.